spec/src/main/asciidoc/query-language.asciidoc

```diff
@@ -0,0 +1,341 @@
+
+==== From clause
+
+The `from` clause, with syntax given by `from_clause`, specifies an _entity name_ which identifies the queried entity. Path expressions occurring in later clauses are interpreted with respect to this entity. That is, the first element of each path expression in the query must be a persistent field of the entity named in the `from` clause. The entity name is a Java identifier, usually the unqualified name of the entity class, as specified in <<Entity Names>>.
+
+The `from` clause is optional in `select` statements. When it is missing, the queried entity is determined by the return type of the repository method, or, if the return type is not an entity type, by the primary entity type of the repository.
+
+NOTE: The syntax of the `update` statement is irregular, with the `from` keyword implied. That is, the syntax _should_ be `update from Entity`, but for historical reasons it is simply `update Entity`.
+
+==== Where clause
+
+The `where` clause, with syntax given by `where_clause`, specifies a conditional expression used to restrict the records returned, deleted, or updated by the query. Only records for which the conditional expression is satisfied are returned, deleted, or updated.
+
+The `where` clause is always optional. When it is missing, there is no restriction, and, and all records are of the queried entity type are returned, deleted, or updated.
+
+==== Select clause
+
+The `select` clause, with syntax given by `select_clause`, specifies a list of expressions which are returned by the query. Each expression in the list is evaluated for each record which satisfies the restriction imposed by the `where` clause. Alternatively, the `select` clause may contain a single `count(*)` aggregate expression, which evaluates to the number of records which satisfy the restriction.
+
+The `select` clause is optional in `select` statements. When it is missing, the query returns the queried entity.
+
+==== Set clause
+
+The `set` clause, with syntax given by `set_clause`, specifies a list of updates to fields of the queried entity. For each record which satisfies the restriction imposed by the `where` clause, and for each element of the list, the scalar expression is evaluated and assigned to the entity field identified by the path expression.
+
+==== Order clause
+
+The `order` clause (or `order by` clause), with syntax given by `orderby_clause`, specifies a lexicographic order for the query results, that is, a list of entity fields used to sort the records which satisfy the restriction imposed by the `where` clause. The keywords `asc` and `desc` specify that a given field should be sorted in ascending or descending order respectively; when neither is specified, ascending order is the default.
+
+The `order` clause is always optional. When it is missing, and when no sort criteria are given as arguments to a parameter of the repository method, the order of the query results is undefined, and may not be deterministic.
```

Comment:
what happens when the db doesn't support `order`?

Reply:
Jakarta Data already has lots of ways to request sorting, so whatever happens in those other cases. Not sure where that's specified.

Reply:
We don't have a guarantee for all fields in NoSQL; for example, Cassandra, when it works, will be for the primary key.
HBase does not support sorting at all; this also happens with key-value databases in the majority of scenarios.

At NoSQL, when the NoSQL database does not support I would expect the unsupportedoperationexception on those scenarios.


I mean, not only for Sorting, but even to operator.

For example, Cassandra has not support to **Or** operator.

Reply:
@graemerocher @otaviojava please review d127e7731db03a209101278ace53c2cf13933d41

Reply:
Another wrinkle here is that the apparently providers must raise `MappingException` when they can't do key-based pagination. That's not intuitive to me at all. It doesn't seem like a `MappingException` from what I can see.

Reply:
> Another wrinkle here is that the apparently providers must raise `MappingException` when they can't do key-based pagination. That's not intuitive to me at all. It doesn't seem like a `MappingException` from what I can see.

We can change that to `UnsupportedOperationException`, too.

Reply:
Please see a4157bf2c0b9b0add9705792da069be07c0154e3.

This should actually be `DataException`, as per our recent discussion.